### ISSUE TITLE
z.lua seems to be broken.

### ISSUE THREAD

**Issue by mcoret** [chrisant996/clink#30]
It seems that the new Lua API are incompatible with [z.lua](https://github.com/skywind3000/z.lua). For instance, it completely ignores "_ZL_ADD_ONCE" environment variable. I've tried, unsuccessfully, to modify the z.lua code and insert the new APIs definitions to the "initialize clink hooks" section, but I don't think I have enough Lua skills to make it properly work.

Thank you for your great work!

**Comment by chrisant996** [comment]
Very interesting, thanks for reporting this.

From the description, offhand I'm not sure what is happening.

I'll take a look.  Maybe there's another v0.4.8 backward compatibility shim needed, or maybe it's something else entirely.

**Comment by chrisant996** [comment]
As far as I can tell, it's working with Clink v1.1.9 for me.

Can you describe specifically what isn't working, and how I can reproduce the problems?
For example, I've heard about z before but I've never used it until now (and I didn't know it integrated with Clink), and I don't know anything about its _ZL_ADD_ONCE env var or how to tell the difference between it ignoring vs respecting the variable.

When I run `z` and `z -h` and `z clink`, they all seem to work fine for me using v1.1.9, so I'm not sure what I should be looking for in order to recognize that it's "incompatible".

Thanks!

**Comment by mcoret** [comment]
For example, z sorts all the (visited) directories according to their rankings which is written in .zlua file (by default in %HOME%), and it's possible to see all the ranking by running "z" w/o arguments.

Each time you're "cd" into a specific directory, its ranking increases. Moreover, each command you're running in/from that specific directory - also increases it's ranking. _ZL_ADD_ONCE variable ensures that the ranking increases only once - when you're "cd"-ing into that directory.

I have Clink v1.1.9.48a3d9, and just by running repeatedly "z" - the directory ranking goes up, irrespectively of _ZL_ADD_ONCE variable.

Thank you again!

**Comment by chrisant996** [comment]
Oh, I see, thanks.

This isn't an API issue.
For the moment, the _ZL_ADD_ONCE variable isn't going to work with Clink v1.0.0 and higher.

Here's what's going on:

Clink v0.4.8 used to only load Lua scripts at startup.
Clink v1.0.0 and higher load Lua scripts at every prompt.
So previously the z.lua script was able to keep some global state information, and now the global state gets discarded at every prompt.

There isn't any information about Martin's reasoning for changing that behavior, but I expect it was to solve reliability problems where scripts could get into strange states due to lingering global variable states.

This was a huge change in the whole structure of the Clink code, and before I invest in rewriting that I need to assess what other side effects might be introduced by rewriting it.

**Comment by chrisant996** [comment]
I'll keep track here of pros/cons of reloading scripts at every prompt:

**Benefits of reloading scripts at every prompt:**
- Scripts naturally automatically reflect changes in Clink settings.  _Can be mitigated by having a way for scripts to hook the beginning of a new prompt._
- Lingering global state is never an issue; if a script misses resetting things properly, there are no negative side effects.

**Drawbacks of reloading scripts at every prompt:**
- Potential performance degradation, if there are many scripts or if they consume significant computation time.
- Scripts cannot have global state that persists across prompts.

**Comment by chrisant996** [comment]
I've changed Clink v1.1.11 to load Lua scripts only once per session.  A new `lua.reload_scripts` setting optionally reloads them each time the edit prompt is activated.  Clink will also automatically reload scripts if the scripts locations change, if the `lua.reload_scripts` setting changes, or if the `clink-reload` command is invoked (<kbd>Ctrl</kbd>+<kbd>x</kbd>,<kbd>Ctrl</kbd>+<kbd>r</kbd>).

I believe the issue is resolved now.  Please reopen it or file a new issue if there are further compatibility issues with z.lua.  Thank you!

<br/>
Along the way, I noticed there's a related bug in [z.lua](https://github.com/skywind3000/z.lua):  changing the _ZL_ADD_ONCE environment variable has no effect until starting a new Clink session or until forcing Lua scripts to reload.  If you want to open an issue in the z.lua repo about that, here is some info about the issue and a potential fix:

It's because of this code in z.lua:
```lua
function z_clink_init()
	local once = os.environ("_ZL_ADD_ONCE", false)
	local previous = ''
	function z_add_to_database()
		pwd = clink.get_cwd()
		if once then
```

It looks like the bug could be fixed by deleting the `local once ...` line and changing the `if once then` line, like this:
```lua
function z_clink_init()
	local previous = ''
	function z_add_to_database()
		pwd = clink.get_cwd()
		if os.environ("_ZL_ADD_ONCE", false) then
```

**Comment by mcoret** [comment]
It seems to work correctly now. I really appreciate it, thank you!

**Comment by mcoret** [comment]
Hi!
I think it's broken again. The history file ".zlua" doesn't get any new updates, at least for the past few days.

**Comment by chrisant996** [comment]
> I think it's broken again. The history file ".zlua" doesn't get any new updates, at least for the past few days.

Can you please add some clarifying information?
- Where is the .zlua file located?
- How is it determined that it isn't updating .zlua?  (e.g. timestamp doesn't change, ratios don't change, something else, etc)
- How is z.lua configured on the machine where it isn't updating?  (e.g. any config changes compared to branch new install)

I'm using  z.lua 1.8.9 (and am about to update to 1.8.11).
My .zlua file is at `c:\users\chrisant\.zlua`.
The timestamp reported by `dir c:\users\chrisant` is updated every time I start a new Clink session or visit a new directory.
The ratios are getting updated as I visit directories.
I have `_ZL_ADD_ONCE=1` and `_ZL_LUA_EXE=c:\luabin\lua52.exe`.

So I'm not able to reproduce a problem where .zlua isn't getting updated.
Can you share additional information to help figure out what's going on?

**Comment by mcoret** [comment]
I think the problem came from my promptfilter: I had a single promptfilter which returns "false" as a second argument. Probably it stopped the further filtering by z. I can't remember changing this return value in the past days (but I can't rule this out completely).
Now, after removing the false, it seems .zlua is getting populated again.

Thank you for the quick response, it helped me realize that the problem is somewhere on my side.

**Comment by chrisant996** [comment]
Ah, sure. Or I would say that's an issue in z.lua itself, really.

You could use the _ZL_CLINK_PROMPT_PRIORITY to change the order in which z's prompt filter gets called.

Ideally z.lua would be updated to use the new `clink.onbeginedit` event instead of a prompt filter (only in new versions of Clink, of course, to stay backward compatible). I might send a pull request for that.

Thank you for reporting the issue, and for the follow-up on it! I'm glad you were able to resolve the issue.
